device/sycl/src/utils/barrier.hpp

```diff
@@ -8,40 +8,59 @@
 #pragma once
 
 // Project include(s).
-#include "traccc/definitions/qualifiers.hpp"
+#include "traccc/device/concepts/barrier.hpp"
 
-// SYCL includes
+// SYCL include(s).
 #include <sycl/sycl.hpp>
 
-namespace traccc::sycl {
+namespace traccc::sycl::details {
 
+/// A barrier type for SYCL kernels.
+template <std::size_t DIMENSIONS = 1>
 struct barrier {
-    barrier(::sycl::nd_item<1> item) : m_item(item) {}
 
-    TRACCC_DEVICE
-    void blockBarrier() { m_item.barrier(); }
+    /// Verify that the barrier is used on a supported dimensionality.
+    static_assert(DIMENSIONS > 0 && DIMENSIONS < 4,
+                  "Only 1D, 2D, and 3D barriers are supported.");
 
-    TRACCC_DEVICE
-    bool blockAnd(bool predicate) {
-        m_item.barrier();
+    /// Dimensions of the barrier
+    static constexpr std::size_t dimensions = DIMENSIONS;
+
+    /// Constructor
+    ///
+    /// @param item The @c ::sycl::nd_item<1> to use for the barrier.
+    ///
+    explicit barrier(const ::sycl::nd_item<dimensions>& item) : m_item(item) {}
```

Comment:
Should we add some template deduction guides? That would be a lot more elegant than having the default template argument.

Reply:
I thought about this. :thinking: But having never written code like that before, wasn't sure how to do it exactly.

You're right though. It would be very cool to construct the correct type using:

```c++
const ::sycl::nd_item<2> item = ...;
traccc::sycl::details::barrier b{item};
```

Any guidelines on how to do this? (To save me some google-ing.)

Reply:
Should be as simple as:

```cpp
template<std::size_t N>
barrier(::sycl::nd_item<N>) -> barrier<N>;
```

And can be placed just behind the class.